src/automata.jl

```diff
@@ -102,50 +120,76 @@ constitute `AbstractAutomaton` interface:
  * `walk`: traveling through the automaton according to given path and initial state
 
 """
-abstract type AbstractAutomaton end
+abstract type AbstractAutomaton{T, N} end
 
 """
     Automaton{T} <: AbstractAutomaton
 Automaton as a vector of states together with alphabet.
 """
-struct Automaton{T} <: AbstractAutomaton
-    states::Vector{State{T}}
+struct Automaton{T, N} <: AbstractAutomaton{T, N}
+    states::Vector{State{T, N}}
     abt::Alphabet
+    uniquenoedge::State{T, N}
 end
 
-# setting the default type to UInt16
-Automaton(states::Vector{Union{State{T}, AbstractState{T}}}, abt::Alphabet) where {T<:Integer} = Automaton{UInt16}(states, abt)
-
-Automaton(abt::Alphabet) = Automaton([State(Word(Int[]), length(abt))], abt)
+# Default type is set to UInt16
+function Automaton(abt::Alphabet)
+    uniquenoedge = State{UInt16, length(abt)}()
+    Automaton{UInt16, length(abt)}([State(Word(), length(abt), uniquenoedge)], abt, uniquenoedge)
+end
 
 states(a::Automaton) = a.states
 initialstate(a::Automaton) = states(a)[1]
+noedge(a::Automaton) = a.uniquenoedge
+
+Base.push!(a::Automaton{T}, s::State{T}) where {T} = push!(a.states, s)
+Base.push!(a::Automaton{T}, name::AbstractWord{T}) where {T} = push!(states(a), State(name, length(a.abt), noedge(a)))
 
-function Base.push!(a::Automaton{T}, s::State{T}) where {T}
-    @assert (length(s.ined) == length(a)) && (length(s.outed) == length(a)) "Tables of in and out edges must have the same length as alphabet"
-    push!(a.states, s)
+"""
+    replace(k::NTuple{N, T}, val, idx::Integer) where {N, T}
+Returns a copy of `NTuple` `k` with value at index `idx` changed to `val`.
+"""
+function replace(k::NTuple{N, T}, val, idx::Integer) where {N, T}
+    @boundscheck 1 ≤ idx ≤ N || throw(BoundsError("Tried to access $(typeof(k)) at index $idx"))
+    return ntuple( i->i==idx ? val : k[i], N)
 end
 
-Base.push!(a::Automaton{T}, name::AbstractWord{T}) where {T} = push!(states(a), State(name, length(a.abt)))
 
+"""
+    updateoutedges!(s::State, to::State, label::Integer)
+Updates outedges of the state `s` with state `to` connected through the edge `label`.
+"""
+function updateoutedges!(s::State, to::State, label::Integer)
+    s.outed = replace(outedges(s), to, label)
+end
+
+"""
+    addinedge!(s::State, from::State)
+Updates inedges of the state `s` with state `from` connected through the edge `label`.
+"""
+function addinedge!(s::State, from::State)
+    push!(inedges(s), from)
+end
 
 """
     addedge!(a::Automaton, label::Integer, from::Integer, to::Integer)
-Adds the edge with a given `label` `from` one state `to` another .
+Adds the edge with a given `label` directed `from` state `to` state.
 """
 function addedge!(a::Automaton, label::Integer, from::Integer, to::Integer)
```

Comment:
* add `Base.@propagate_inbounds` and replace `@asserts` by `@boundscheck checkbounds(a.abt, label)`, etc
* prepend `@inbounds` to all lines 182-185. we have checked above that these should be safe. In general we should make sure that all of those operations are `@inbounds` after the check.

also: `length(a.abt)` → `N`